### PR TITLE
Convert plane roll/pitch PIDs to use AC_PID

### PR DESCRIPTION
**Pull request by tridge**
This converts plane to using AC_PID for roll and pitch.

Moving to AC_PID gives us a much more standard PID structure, which will allow for more uniform PID tuning guidance for users and a common automatic tuning system. It also gives us PID element filters from AC_PID which are very useful for controlling noise

This also implements a new fixed-wing autotune system that can estimate FF, P, I and D. The new autotune should be much more robust than the previous algorithm

### CODE REVIEW

File: libraries/APM_Control/AP_AutoTune.cpp [hunk 1]
```diff
@@ -211,24 +208,22 @@ void AP_AutoTune::check_state_exit(uint32_t state_time_ms)
         // we increase P if we have not saturated the surfaces during
```
Comment (khancyr) [inline libraries/APM_Control/AP_AutoTune.cpp:208]: comment is wrong, isn't it ?

**Review by priseborough — COMMENTED** [review]
I've been through the maths for the gain conversions and the loops and other than the addition of the rate derivative gain, the control loop transfer function should be is equivalent to what we had before. We will need a follow up PR to fix https://ardupilot.org/plane/docs/roll-pitch-controller-tuning.html as the diagrams and tuning descriptions are now out of date.

**Comment by tridge** [comment]
> I will follow up with some flight testing.

thanks Paul!

**Comment by tridge** [comment]
@priseborough one thing I'm concerned about is we can end up with quite low P values in autotune, and that leads to small I values as we set I=0.75*P. The P can end up very small if SMAX is too low.
This leads to problems in hand launches with small CoG errors. The trim may not be learnt quickly enough by the integrator to prevent hitting the ground.
I'm thinking we could have a minimum I value corresponding to learning a 25% trim of the surface in 3 seconds with a rate error equal to the min of RMAX and attitude_limit/tconst

**Comment by priseborough** [comment]
No issues with flight tests.

**Review by priseborough — APPROVED** [review]

**Comment by tridge** [comment]
@priseborough thanks Paul! I think we need to either prevent getting very small P values, or improve integrator handling, before this is usable for autotune. The size of P/I we're getting by default in autotune is too small

**Comment by priseborough** [comment]
Looking at the SlewLimiter::modifier function as used by the plane auto-tune, the gust sensitivity could be reduced by checking for a negative and positive slew event. I've pushed a patch to a branch in my repo here which illustrates the algorithm. 

https://github.com/priseborough/ardupilot/commit/3c9d25eb8e227ee82220f548390378c151f860af

In the process of doing this I noticed that the SlewLimiter::modifier function was storing a modified value of the last_sample = mod * sample. I think it should have been the unmodified sample given the way the function is being used. 

Initial flight testing of the autotune function on a Nano Goblin <250g airframe indicated that the FF gain was being underestimated likely due to gust activity as evidenced by the achieved vs demanded roll and the roll PID integrator integrating in the direction of roll during turn reversals , eg:

Screen Shot 2021-04-06 at 8 32 51 pm
Screen Shot 2021-04-06 at 8 32 30 pm
Screen Shot 2021-04-06 at 8 31 33 pm

The filter that is applied to the actuator and body rate used for the FF calculation will have to be slowed down. the value of 2Hz used correspinds to a circular frequency of 2 * 2 * pi rad/sec which is too high given that for a first order filter, the time constant = 1 / circ_freq. A value between 0.5 and 1.0 Hz would be better.

**Comment by Hwurzburg** [comment]
@tridge @priseborough so here is the log showing failure of Dmod to stop autotakeoff oscillations....note no real DMod the rest of the flight so the tune is tight but okay...Dmod is .2 but since airspeed estimate pegs the scaler initially it still oscillates... as we discussed, perhaps a variant of my PR #17068 which limits scaling to <1 during autotakeoffs to prevent headwind from lowering airspeed estimates, which also qualifies it based on heavy Dmod, so that the scaling is not reduced if not detecting oscillations would be better..  https://www.dropbox.com/s/sugk43d66geyo4f/dart_no_fix_4_5.bin?dl=0

**Comment by tridge** [comment]
@priseborough thanks Paul! I've added your changes. 
Testing in RF9 seems to show that with the slew limiter change that we need a much lower SMAX value to effectively suppress oscillations. Is that expected?

**Comment by priseborough** [comment]
A lower value is expected, a much lower value is not. I'll check it out.

**Comment by Hwurzburg** [comment]
@tridge @priseborough  Tridge asked that I attach this log https://www.dropbox.com/s/sugk43d66geyo4f/dart_no_fix_4_5.bin?dl=0  which shows that Dmod was not stopping oscillation on a properly tuned airspeed less autotakeoff....although Pitch had no Dmod hits to speak of the rest of the flight,indicating a non-marginal tune,  it could not overcome high scaling gain while it was climbing into a headwind...it did not do it the previous day when using #17068....which Tridge thinks should be made a method and qualified with low Dmod, ie only when we think its oscillating to reduce scaling gain to allow for overspeed estimates also...log was done 3 days ago with this PR

**Comment by tridge** [comment]
> fyi, this PR rebased on master no longer builds on 1MB boards

it certainly builds on some boards as that is tested in CI. Which board did you check?

**Comment by Hwurzburg** [comment]
> > fyi, this PR rebased on master no longer builds on 1MB boards
> 
> it certainly builds on some boards as that is tested in CI. Which board did you check?

overflows by 312bytes on omnibusf4pro if I just fetch this PR direcly and build it....no rebasing....on 4/4, it built fine, even rebased on master...master builds omnibus4pro with 3.4K to spare

**Comment by tridge** [comment]
@priseborough example log http://uav.tridgell.net/tmp/00000046.BIN
I needed a lower SMAX to prevent oscillation with new slew limiter

**Comment by priseborough** [comment]
@tridge I've reworked the slew rate limiter function to require a minimum number of crossing detections. See https://github.com/priseborough/ardupilot/commit/ee80b1f8fd0bd303a58ee9dd2e255f0bcfba3385

It appears to work OK on SITL with reasonable slew rate level output. I may end up adding a filter to the output slew rate to smooth it out further.

**Comment by priseborough** [comment]
I've added a filter to the increase of the slew rate metric (application of of compression) here: https://github.com/priseborough/ardupilot/commit/1f19cb627e1380b4d12548ad4e15343c0e3e9035

Here's some results fromSITL showing the effect of increasing RLL_RATE_P from the default of 0.3 to 2.5, then 3.0, then 2.0, then back to 0.3. FBWB mode was then selected and an alternating series of left and right full bank steps demanded.

Screen Shot 2021-04-10 at 9 55 26 pm

Screen Shot 2021-04-10 at 9 55 45 pm

**Comment by priseborough** [comment]
I've made some further refinements to the SlewLimiter function and pushed them to my version of this branch: https://github.com/priseborough/ardupilot/commits/pr-plane-ac-pid

I was able to do a auto-tune successfully in SITL with reasonable results after splitting the filtering strategy for the externally reported slew rate used by the autotune vs the one used for the PID gain compression due to the differing requirements of these algorithms.

Screen Shot 2021-04-11 at 9 44 27 am
Screen Shot 2021-04-11 at 9 43 54 am
Screen Shot 2021-04-11 at 9 43 48 am
Screen Shot 2021-04-11 at 9 43 42 am
Screen Shot 2021-04-11 at 9 43 37 am

**Comment by priseborough** [comment]
Closer inspection of the SITL autotune log generated by https://github.com/priseborough/ardupilot/commits/pr-plane-ac-pid at https://github.com/priseborough/ardupilot/commit/827ba35d4fdabc3667da9e323154684462729d59 shows that the P and D gains are reducing until the lightly damped oscillation following a roll manoeuvre disappears, eg:

Screen Shot 2021-04-11 at 10 14 59 am
Screen Shot 2021-04-11 at 10 14 49 am

**Comment by Hwurzburg** [comment]
I think that something might need to be updated regarding the @SYS logging also....getting failures when I try to access the files when using this firmware...

**Pull request merged**